File: src/C3.Catalogue/Tapes/TapeOperationResult.vb
Namespace Tapes

    Public Enum TapeFailure
        None = 0
        ModelRequired
        ModelNotFound
        InvalidBulkCount
        IdentifierCapacityExceeded
        SideNameRequired
        DuplicateIdentifier
        NotFound
        StorageFailure
    End Enum

    Public NotInheritable Class TapeOperationResult

        Private Sub New()
            Tapes = New List(Of Tape)()
        End Sub

        Public Property IsSuccess As Boolean
        Public Property Tapes As IList(Of Tape)
        Public Property Failure As TapeFailure
        Public Property Message As String

        Public Shared Function Success(values As IList(Of Tape)) As TapeOperationResult
            Return New TapeOperationResult() With {
                .IsSuccess = True,
                .Tapes = values,
                .Failure = TapeFailure.None,
                .Message = String.Empty
            }
        End Function

        Public Shared Function Failed(failure As TapeFailure, message As String) As TapeOperationResult
            Return New TapeOperationResult() With {
                .IsSuccess = False,
                .Failure = failure,
                .Message = message
            }
        End Function

    End Class

End Namespace

File: src/C3.Catalogue/Tapes/TapeOperationResult.cs
using System.Collections.Generic;

namespace C3.Catalogue.Tapes
{
    public enum TapeFailure
    {
        None = 0,
        ModelRequired,
        ModelNotFound,
        InvalidBulkCount,
        IdentifierCapacityExceeded,
        SideNameRequired,
        DuplicateIdentifier,
        NotFound,
        StorageFailure
    }

    public sealed class TapeOperationResult
    {
        private TapeOperationResult()
        {
            Tapes = new List<Tape>();
        }

        public bool IsSuccess { get; set; }

        public IList<Tape> Tapes { get; set; }

        public TapeFailure Failure { get; set; }

        public string Message { get; set; }

        public static TapeOperationResult Success(IList<Tape> values)
        {
            return new TapeOperationResult
            {
                IsSuccess = true,
                Tapes = values,
                Failure = TapeFailure.None,
                Message = string.Empty
            };
        }

        public static TapeOperationResult Failed(TapeFailure failure, string message)
        {
            return new TapeOperationResult
            {
                IsSuccess = false,
                Failure = failure,
                Message = message
            };
        }
    }
}
